From ff7b0a3539654566c13a8cb8aa97a9832431133a Mon Sep 17 00:00:00 2001
From: Ritika Chopra <r.chopra@dozenos.local>
Date: Mon, 5 Oct 2026 10:16:26 -0700
Subject: [PATCH 2/2] linux-cp: T9243: preserve timestamp ordering for
 dump-imported VLANs

VLAN LIPs created from RTM_GETLINK dump replies have no per-message
timestamp context. Record a cutoff immediately before sending the link
dump request and use it as the creation timestamp for LIPs created from
that dump.

This ensures link notifications received after the dump request have a
newer timestamp than the imported LIP and are not incorrectly rejected
as early messages.
---
 src/plugins/linux-cp/lcp_nl.c     | 44 ++++++++++++++++++++++++++++---
 src/plugins/linux-cp/lcp_nl.h     |  2 ++
 src/plugins/linux-cp/lcp_router.c | 11 +++++---
 3 files changed, 51 insertions(+), 6 deletions(-)

diff --git a/src/plugins/linux-cp/lcp_nl.c b/src/plugins/linux-cp/lcp_nl.c
index 9c08ac299..da8edd48a 100644
--- a/src/plugins/linux-cp/lcp_nl.c
+++ b/src/plugins/linux-cp/lcp_nl.c
@@ -70,6 +70,12 @@ typedef enum nl_event_type_t_
   NL_EVENT_SYNC
 } nl_event_type_t;
 
+typedef struct nl_link_dump_t_
+{
+  f64 ts;
+  u8 active;
+} nl_link_dump_t;
+
 typedef struct nl_main
 {
 
@@ -91,6 +97,7 @@ typedef struct nl_main
   u32 sync_batch_limit;
   u32 sync_batch_delay_ms;
   u32 sync_attempt_delay_ms;
+  nl_link_dump_t link_dump;
 
   volatile bool is_syncing;
 
@@ -113,6 +120,10 @@ static nl_main_t nl_main = {
   .sync_batch_limit = NL_SYNC_BATCH_LIMIT_DEF,
   .sync_batch_delay_ms = NL_SYNC_BATCH_DELAY_MS_DEF,
   .sync_attempt_delay_ms = NL_SYNC_ATTEMPT_DELAY_MS_DEF,
+  .link_dump = {
+    .ts = 0,
+    .active = 0,
+  },
 };
 
 /* #define foreach_nl_nft_proto  \ */
@@ -405,10 +416,17 @@ lcp_nl_route_send_dump_req (nl_sock_type_t sock_type, int msg_type)
   nl_main_t *nm = &nl_main;
   struct nl_sock *sk_route = nm->sk_route_sync[sock_type];
   int err;
+  f64 dump_ts = 0;
   struct rtgenmsg rt_hdr = {
     .rtgen_family = AF_UNSPEC,
   };
 
+  if (sock_type == NL_SOCK_TYPE_LINK)
+    {
+      nm->link_dump.active = 0;
+      dump_ts = vlib_time_now (vlib_get_main ());
+    }
+
   err =
     nl_send_simple (sk_route, msg_type, NLM_F_DUMP, &rt_hdr, sizeof (rt_hdr));
 
@@ -417,12 +435,29 @@ lcp_nl_route_send_dump_req (nl_sock_type_t sock_type, int msg_type)
       NL_ERROR ("Unable to send a dump request: %s", nl_geterror (err));
     }
   else
-    NL_INFO ("Dump request sent via socket %d of type %d",
-	     nl_socket_get_fd (sk_route), sock_type);
+    {
+      if (sock_type == NL_SOCK_TYPE_LINK)
+	{
+	  nm->link_dump.ts = dump_ts;
+	  nm->link_dump.active = 1;
+        }
+      NL_INFO ("Dump request sent via socket %d of type %d",
+	       nl_socket_get_fd (sk_route), sock_type);
+    }
 
   return err;
 }
 
+u8
+lcp_nl_get_link_dump_ts (f64 *ts)
+{
+  if (!nl_main.link_dump.active || !ts)
+    return 0;
+
+  *ts = nl_main.link_dump.ts;
+  return 1;
+}
+
 static int
 lcp_nl_route_dump_cb (struct nl_msg *msg, void *arg)
 {
@@ -694,7 +729,9 @@ nl_route_process (vlib_main_t *vm, vlib_node_runtime_t *node,
 	goto sync_later;                                                      \
     }                                                                         \
   while (!is_done);                                                           \
-  nl_##fn##_sync_end ();
+  nl_##fn##_sync_end ();                                                      \
+  if (stype == NL_SOCK_TYPE_LINK)                                            \
+    nm->link_dump.active = 0;
 
 	    foreach_sock_type
 #undef _
@@ -708,6 +745,7 @@ nl_route_process (vlib_main_t *vm, vlib_node_runtime_t *node,
 	  wait_time = (vec_len (nm->nl_msg_queue) != 0) ? 1e-3 : DAY_F64;
 
 	sync_later:
+	  nm->link_dump.active = 0;
 	  /* Close netlink synchronization sockets */
 #define _(stype, mtype, tname, fn) lcp_nl_close_sync_socket (stype);
 	  foreach_sock_type
diff --git a/src/plugins/linux-cp/lcp_nl.h b/src/plugins/linux-cp/lcp_nl.h
index 1b0188d7e..09aa19656 100644
--- a/src/plugins/linux-cp/lcp_nl.h
+++ b/src/plugins/linux-cp/lcp_nl.h
@@ -153,6 +153,8 @@ typedef struct nl_msg_info
 
 extern struct nl_cache *lcp_nl_get_cache (lcp_nl_obj_t t);
 extern int lcp_nl_drain_messages (void);
+/* Returns non-zero only while link-dump replies are being dispatched. */
+extern u8 lcp_nl_get_link_dump_ts (f64 *ts);
 extern void lcp_nl_set_buffer_size (u32 buf_size);
 extern void lcp_nl_set_batch_size (u32 batch_size);
 extern void lcp_nl_set_batch_delay (u32 batch_delay_ms);
diff --git a/src/plugins/linux-cp/lcp_router.c b/src/plugins/linux-cp/lcp_router.c
index 2652da2cd..01718f1c7 100644
--- a/src/plugins/linux-cp/lcp_router.c
+++ b/src/plugins/linux-cp/lcp_router.c
@@ -423,6 +423,7 @@ lcp_router_link_add (struct rtnl_link *rl, void *ctx)
 	  lcp_itf_pair_t *sub_lip;
 	  int vlan;
 	  int rv;
+	  f64 dump_ts;
 	  u8 *ns = 0; /* FIXME */
 
 	  lip = lcp_itf_pair_get (lipi);
@@ -480,9 +481,13 @@ lcp_router_link_add (struct rtnl_link *rl, void *ctx)
 
 	  sub_lip = lcp_itf_pair_get (
 	    lcp_itf_pair_find_by_phy (sub_phy_sw_if_index));
-	  if (sub_lip && ctx)
-	    sub_lip->lip_create_ts =
-	      ((nl_msg_info_t *) ctx)->ts;
+	  if (sub_lip)
+	    {
+	      if (ctx)
+	        sub_lip->lip_create_ts = ((nl_msg_info_t *) ctx)->ts;
+	      else if (lcp_nl_get_link_dump_ts (&dump_ts))
+	        sub_lip->lip_create_ts = dump_ts;
+	    }
 
 	  if (up)
 	    vnet_sw_interface_admin_up (vnet_get_main (), sub_phy_sw_if_index);
-- 
2.51.2

